modules/apps/frontend-js/frontend-js-components-web/src/main/resources/META-INF/resources/translation_manager/TranslationAdminSelector.js

```diff
@@ -0,0 +1,220 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import ClayButton from '@clayui/button';
+import ClayDropDown from '@clayui/drop-down';
+import ClayIcon from '@clayui/icon';
+import ClayLabel from '@clayui/label';
+import ClayLayout from '@clayui/layout';
+import PropTypes from 'prop-types';
+import React, {useEffect, useMemo, useState} from 'react';
+
+import TranslationAdminModal from './TranslationAdminModal';
+
+const emptyArray = [];
+
+const TranslationAdminSelector = ({
+	activeLocales: initialActiveLocalesIds = emptyArray,
+	adminMode,
+	ariaLabels = {
+		default: Liferay.Language.get('default'),
+		manageTranslations: Liferay.Language.get('manage-translations'),
+		translated: Liferay.Language.get('translated'),
+		untranslated: Liferay.Language.get('untranslated'),
+	},
+	availableLocales = emptyArray,
+	defaultLocaleId,
+	selectedLocaleId: initialSelectedLocaleId,
+	showOnlyFlags,
+	small = false,
+	translations,
+}) => {
+	const [activeLocalesIds, setActiveLocalesIds] = useState(
+		initialActiveLocalesIds
+	);
+	const [selectedLocaleId, setSelectedLocaleId] = useState(
+		initialSelectedLocaleId
+	);
+	const [selectorDropdownActive, setSelectorDropdownActive] = useState(false);
+	const [translationModalVisible, setTranslationModalVisible] = useState(
+		false
+	);
+
+	const handleCloseTranslationModal = (activeLocalesIds) => {
+		setActiveLocalesIds(activeLocalesIds);
+		setTranslationModalVisible(false);
+	};
+
+	const activeLocales = useMemo(
+		() =>
+			availableLocales.filter((availableLocale) =>
+				activeLocalesIds.includes(availableLocale.id)
+			),
+		[availableLocales, activeLocalesIds]
+	);
+
+	const selectedLocale = useMemo(() => {
+		const id = selectedLocaleId ?? defaultLocaleId;
+
+		return availableLocales.find(
+			(availableLocale) => availableLocale.id === id
+		);
+	}, [availableLocales, defaultLocaleId, selectedLocaleId]);
+
+	useEffect(() => {
+		setActiveLocalesIds(initialActiveLocalesIds);
+	}, [initialActiveLocalesIds]);
```

Comment:
This is, as you said, a controlled-uncontrolled component, indeed.

We want the user to be able to pass `activeLocaleIds` as a param every time the want to change it, but we're handling those activeLocales internally. That's why we're renaming the prop to `initialActiveLocaleIds`, and because we don't feel that exposing an API with the name `initialWhatever` is a good naming strategy (since you change it at any moment so it won't be initial at all).

That's why we use this pattern:

1. Rename the prop to `initialWhatever`
2. Create a state with the original a name, setting the initialValue with the received prop
3. Create a useEffect to update the state with the new param received.

Reply:
About adding `onWhateverChanges` API I prefer not to add them until someone need them. Keep it simple.

Reply:
@markocikos nope, it's not enough with the initial value of useState since that initial value is only set when the component is created, so if the received prop value changes you need to listen to it with a useEffect.

Reply:
> You can find a sample here: https://codesandbox.io/s/empty-star-0yxcw?file=/src/App.js

In the sample, change 
```js
<button onClick={() => setMyProp("value2")}>Click</button>
```
to
```js
<button onClick={() => setMyProp(Math.random())}>Click</button>
```
and you'll see `myProp` value changes. It is controlled by an external component (`App`). If external component sets prop repeatedly to `value2`, it will not change and re-render.

Reply:
mmmm, nope, is not changing without the `useEffect` (refresh the sample before clicking the button)

Reply:
https://codesandbox.io/s/blissful-dawn-xhim1?file=/src/App.js

Reply:
In that sample you're printing the prop, not the state

Reply:
Ah, I see the problem. Can we pass the setter from controlling component? That would be a cleaner design.

https://codesandbox.io/s/objective-glitter-z7stg?file=/src/App.js

Reply:
> About adding onWhateverChanges API I prefer not to add them until someone need them. Keep it simple.

I'm not sure if this would be a decision to keep it simple but generate strange behavior and it's also counterintuitive to implement in React, for example, dealing with controlled and uncontrolled components at the same time is bad practice if you manipulate an element directly native in React, this throws errors on the console, I'm not saying this is the case but what I see is a little strange, for example:

The component controls the state internally and also allows the developer to manipulate the state, that's not wrong but I think the way it was implemented is weird and it's not very intuitive even for the developer.

```jsx
// Since the `activeLocales` is being manipulated internally the `activeLocales` from the
// outside have no visibility as to what the status is now. So you have a problem that the
// state is out of sync since you allow to update at any time.
const [activeLocales, setActiveLocales] = useState([]);

<TranslationAdminSelector
  activeLocales={activeLocales}
/>

// Unlike you allow the state to be known from the outside and the inside and both will be
// synced and you won't have problems with any of the states being stale, also the
// developer needs to know if `activeLocales` has changed and `selectedLocaleId` has
// changed as well.
const [activeLocales, setActiveLocales] = useState([]);

<TranslationAdminSelector
  activeLocales={activeLocales}
  onActiveLocales={setActiveLocales}
/>
```

What I mean is that if the developer uses this component he won't have information if the locale was selected or if the `activeLocalesIds` changed, it's important for the application to be able to change the current language or remove the translations because a translation was removed .

For example in Clay, we have `useInternalState` that takes care of this, if the two properties are not passed we use the internal state, otherwise we will use the external state but you cannot have two states and be out of sync.

Reply:
That way you're forcing the user tu use the component as a controlled component, which I don't want here. The component works on his own, but also has the ability to be updated from the parent component if needed. This is a pretty common pattern and I see it as clean as any other pattern

Reply:
Hmm... ok! how can I know if the translation has been added or removed? also, how can I know if a language has been removed or added to the locales list? I'm asking this because it is necessary for the application to control since who changes this information is a component, right?

> That way you're forcing the user tu use the component as a controlled component, which I don't want here.

You don't have to force, it will just be consistent when you want it to be controlled. This code snippet is saying it:

```js
useEffect(() => {
  setActiveLocalesIds(initialActiveLocalesIds);
}, [initialActiveLocalesIds]);
```

I would say it's a common behavior when you don't manipulate that value without the support of a callback to listen for changes to that value. Does it make sense to you?

Reply:
>That way you're forcing the user tu use the component as a controlled component, which I don't want here. The component works on his own, but also has the ability to be updated from the parent component if needed. This is a pretty common pattern and I see it as clean as any other pattern

That comment was a response to Marko's:
>Ah, I see the problem. Can we pass the setter from controlling component? That would be a cleaner design.

Didn't see yours, Matu, when I wrote it (thx github)

Reply:
Oh, no problem 😅

Reply:
About the callback I can add it, it probably makes sense in this case, but not necessarily always does.

Reply:
> About the callback I can add it, it probably makes sense in this case, but not necessarily always does.

Yes, I think some cases really won't be necessary. We can make it optional and, if it's passed to the props, we switch to the controlled state.

Reply:
> as clean as any other pattern

I just don't like `useEffect`. It feels like 99% of issues I came across in React have some weird `useEffect` interaction as the underlying cause. It promotes sloppy programming, where you globally hook onto state change, instead of following natural code flow.

Reply:
![kitten](https://user-images.githubusercontent.com/5435511/131468585-dce97b87-7eb9-4eae-9072-38f5988b4018.jpg)

Reply:
to.me useEffect is part of the natural flow of a React component since React is based on that: reacting to stuff. The problem is not using useEffect, is using it to solve every problem when you have no idea what you're doing and no checkimg what actually happen, which usually causes múltiple renders and unexpected behaviors.

What we should not do is to provide more complex APIs because we don't trust our code 🤷🏻‍♂️

Reply:
By 'natural flow', I meant code with strict scope rules. Function arguments. Local variables. Unique action handlers. Separation of concerns. In hindsight, 'natural flow' is not a good term, it's subjective, I should not have used it.

It's the same as using custom event bus, global event listeners, or simply mutable global variables. Can you do it in JS? Sure. Sometimes you need to. But it doesn't make it a good practice, as it always leads to messy, disjointed and hard to understand code. It's something we should avoid, whenever possible.

Reply:
@carloslancha  Bottom line, both @matuzalemsteles and me are suggesting the same pattern, where control is cleanly kept in the parent component:

```js
const [activeLocales, setActiveLocales] = useState([]);

<TranslationAdminSelector
  activeLocales={activeLocales}
  onActiveLocales={setActiveLocales}
/>
```
or 

https://codesandbox.io/s/objective-glitter-z7stg?file=/src/App.js

Those two are basically the same pattern.

This would remove the need for `setEffect`, and also makesdiscussion about default values in https://github.com/liferay-frontend/liferay-portal/pull/1372#discussion_r697561675 irrelevant. You can keep normal inline values, without making a variable out of it.

Reply:
As I explained before, that way we're forcing the user to handle `activeLocales` state out of the component in a controlled way, which we don't necessarily want.

One use case could be the user listening for `selectedLanguageId` changes but not worrying about the `activeLanguageIds` because they don't needed.

I've just added an `onSelectedLanguageIdChange` btw, which was missing.

Reply:
> As I explained before, that way we're forcing the user to handle activeLocales state out of the component in a controlled way, which we don't necessarily want.

Why not? :thinking:  I don't see how your example with `selectedLanguageId` relates to this. If anything, it should be given the same treatment, passing in `setSelectedLanguageId` as `onSelectedLanguageIdChange`, from parent component.

Reply:
Hmm I think we really shouldn't force the state to be controlled by the Parent only when necessary, for example instead of using `useEffect` to keep up to date use the internal state conditionally if you take into account the implementation with the latest changes, that should look like this.

```js
const [internalActiveLanguageIds, setInternalActiveLanguageIds] = useState(
  initialActiveLanguageIds
);

const activeLanguageIds = initialActiveLanguageIds ? initialActiveLanguageIds : internalActiveLanguageIds;
const setActiveLanguageIds = onActiveLanguageIdsChange ? onActiveLanguageIdsChange : setInternalActiveLanguageIds;
```

This reduces the use of `useEffect` for this case, because you will always sync and have to be reactive to changes which is not always very good in that case you would be choosing your source of truth but also leaving state control optional. Anyway, I don't think this is a blocker for this PR, if we find this valid we can send in another PR.

Reply:
Having `initialActiveLanguageIds` doesn't imply the state is going to be handled by the parent, only that the component will have an initial value, so I don't think that approach will work.

IMHO guys I think we're creating/seeing problems where there's none. `useEffect` is there to help us with side effects handling, which is exactly what we're doing here: if we receive a new value for a prop we update the internal state. That's it. And with that we allow the user to use the component/prop as a controlled or an uncontrolled one.

It's a very simple component and it's very easy to understand IMO.

Btw, I've created some tests just to be sure everything is working (see https://github.com/liferay-frontend/liferay-portal/pull/1425) and I found an issue (when you remove the current selected locale from the list of active locales it were not updating the selected one with the default locale).

Did any of you actually test the feature? We need to test them carefully (me too, of course) to be sure that everything is working and not only focus on the code.

Thx guys!